JOSS Submission (#439)
lincc-frameworks/python-project-template: paper.md

---
title: 'A Python Project Template for Healthy Scientific Software'
tags:
  - Python
  - template
authors:
  - name: Drew Oldag

>     orcid: asdfasdf

Reviewer: ```suggestion
    orcid: 0000-0001-6984-8411
```

    equal-contrib: true
    affiliation: "1, 2"
  - name: Melissa DeLucchi
    orcid: asdfasdf
    equal-contrib: true
    affiliation: "1, 3"
  - name: Wilson Beebe
    orcid: asdfasdf
    affiliation: "1, 2"
  - name: Doug Branton
    orcid: asdfasdf
    affiliation: "1, 2"
  - name: Sandro Campos
    orcid: asdfasdf
    affiliation: "1, 3"
  - name: Carl Christofferson
    orcid: asdfasdf
    affiliation: "1, 2"
  - name: Andrew Connolly <-----------????
    orcid: asdfasdf
    affiliation: "1, 2"
  - name: Jeremy Kubica
    orcid: asdfasdf
    affiliation: "1, 3"
  - name: Olivia Lynn
    orcid: asdfasdf
    affiliation: "1, 3"

>   - name: Kostya Malanchev

Reviewer 2: ```suggestion
  - name: Konstantin Malanchev
```

    orcid: asdfasdf
    affiliation: "1, 3"

>   - name: Alex Malz <----------????

Author: Fairly confident that Alex didn't contribute to the PPT, but please correct me if I'm forgetting or not considering something.

    orcid: asdfasdf
    affiliation: "1, 3"
  - name: Sean McGuire
    orcid: asdfasdf
    affiliation: "1, 3"
  - name: Chris Wenneman
    orcid: asdfasdf
    affiliation: "1, 2"
affiliations:

>   - name: LINCC-Frameworks

Reviewer 3: I'm very excited about LINCC being an affiliation!!

    index: 1

>   - name: University of Washington

Reviewer 3: ```suggestion
  - name: DiRAC Institute and the Department of Astronomy, University of Washington, 3910 15th Ave NE, Seattle, WA 98195, USA
```

    index: 2

>   - name: Carnegie Mellon University

Reviewer 3: ```suggestion
  - name: McWilliams Center for Cosmology, Department of Physics, Carnegie Mellon University, Pittsburgh, PA 15213, USA
```

## Distribution

Software is only useful to the broader community if other users can find, install, and update it. Package management systems such as pip[^3] or conda[^4] provide simple tools for a user to download a new software package and its dependencies from public repositories such as PyPI[^7] or conda-forge[^8]. To enable easy code installation with pip or conda the LF-PPT provides support for automatic distribution through both PyPI and conda-forge when the user applies a new git tag via the GitHub UI.

## Additional code health tools

The LF-PPT supports many more features than those listed here including static code analysis, performance benchmarking, code testing coverage and more. For a complete list, please see the documentation at http://lincc-ppt.readthedocs.io/. 

The most notable exclusion from the LF-PPT is code. Aside from a few optional stub source and test files, the LF-PPT allows the user focus on scientific code, while supporting them with the industry best practices in maintainable software engineering.  

# Usage

## General usage

> The LINCC-Frameworks Python Project Template automates the setup of the above processes for Python projects hosted on GitHub. The only direct dependency is copier[^5] [@copier], which is used as the engine to generate new projects from the LF-PPT with a specific directory structure, the requested configuration and stub files. To hydrate a new or existing project with the LF-PPT via copier the user calls:

Reviewer: Probably worth re-ordering the footnotes to have them in order. We've already used 7 and 8.

Author: Yep, totally makes sense to me, I'll get the ordering sorted.

Author: Done

> > copier copy gh:lincc-frameworks/python-project-template <new/project/directory>

Reviewer 3: Turns out, you can't suggest adding three back-ticks - kinda breaks the github UI.

Anyway. I think we should use backtick-code-blocks for this and the other command line prompts. The ">" character in markdown implies quoted text, and is rendered differently.

Author: Gotcha, I can do that. Thanks!

In addition to creating new projects from scratch, the LF-PPT can be applied to existing projects to incorporate the features it provides. Often a more established project will require more effort to apply the LF-PPT. However, LINCC-Frameworks has had success applying the template to multiple legacy projects. A collection of tips for applying the template to existing projects is included in the documentation http://lincc-ppt.readthedocs.io/.

## Updating projects

As best practices evolve and new tools are introduced, the LF-PPT will incorporate those into the template. Updates to the template can be applied to existing projects with minimal effort, allowing the users to focus on science and not software maintenance.

> copier update

## Hibernating projects

Scientific projects often go into periods of hibernation when not under active development or use, and are often challenging to revive. With the LF-PPT, hibernating projects should be much easier to reactivate. Automatically scheduled smoke tests and dependabot[^6] integration ensure that the code and dependencies continue to function correctly without significant interaction from the maintainers.

## Recent applications

> The LF-PPT has been applied to multiple LINCC Frameworks projects including LSDB[^11], Hipscat[^12], and TAPE[^13], short term collaborations such as Sorcha[14], SuperPhot+[^10], DeepDISC[^15], and MacCauff[^9], and external project including kcorrect[^16] and FlexCode[^17] Additionally a project specific version was forked from the project and has been applied to all RAIL packages[^18].

Reviewer 3: ```suggestion
The LF-PPT has been applied to multiple LINCC Frameworks projects including LSDB[^11], Hipscat[^12], and TAPE[^13], short term collaborations such as Sorcha[^14], SuperPhot+[^10], DeepDISC[^15], and MacCauff[^9], and external project including kcorrect[^16] and FlexCode[^17] Additionally a project specific version was forked from the project and has been applied to all RAIL packages[^18].
```

# Acknowledgments

This project is supported by Schmidt Sciences.

[^1]: https://www.sphinx-doc.org/
[^2]: https://about.readthedocs.com/
[^3]: https://pip.pypa.io/en/latest/
[^4]: https://docs.conda.io/en/latest/
[^5]: https://copier.readthedocs.io/en/stable/ 
[^6]: https://docs.github.com/en/code-security/dependabot 
[^7]: https://pypi.org/
[^8]: https://conda-forge.org/docs/

> [^9]: https://something-for-maccauff???

Reviewer 3: ```suggestion
[^9]: https://github.com/macauff/macauff
```

authors:
  - name: Drew Oldag
    orcid: 0000-0001-6984-8411
    equal-contrib: true
    affiliation: "1, 2"
  - name: Melissa DeLucchi
    orcid: 0000-0002-1074-2900
    equal-contrib: true
    affiliation: "1, 3"
  - name: Wilson Beebe
    orcid: 0009-0003-1791-8707
    affiliation: "1, 2"
  - name: Doug Branton
    orcid: 0009-0009-7822-7110
    affiliation: "1, 2"
  - name: Sandro Campos
    orcid: 0009-0007-9870-9032
    affiliation: "1, 3"
  - name: Carl Christofferson
    orcid: 0009-0009-1164-3494
    affiliation: "1, 2"
  - name: Andrew Connolly
    orcid: 0000-0001-5576-8189
    affiliation: "1, 2"
  - name: Jeremy Kubica
    orcid: 0009-0009-2281-7031
    affiliation: "1, 3"
  - name: Olivia Lynn
    orcid: 0000-0001-5028-146X
    affiliation: "1, 3"
  - name: Konstantin Malanchev
    orcid: 0000-0001-7179-7406
    affiliation: "1, 3"
  - name: Alex Malz
    orcid: 0000-0002-8676-1622
    affiliation: "1, 3"
  - name: Rachel Mandelbaum
    orcid: 0000-0003-2271-1527
    affiliation: "1, 3"
  - name: Sean McGuire
    orcid: 0009-0005-8764-2608
    affiliation: "1, 3"
  - name: Chris Wenneman
    orcid: 0009-0002-6218-5007
    affiliation: "1, 2"
affiliations:

>   - name: LINCC-Frameworks

Reviewer 4: Should not have a hyphen... just "LINCC Frameworks" (here and throughout)

Author: Thanks for pointing this out. I'll fix this up throughout the paper.

Author: Done

    index: 1
  - name: DiRAC Institute and the Department of Astronomy, University of Washington, 3910 15th Ave NE, Seattle, WA 98195, USA
    index: 2
  - name: The McWilliams Center for Cosmology & Astrophysics, Department of Physics, Carnegie Mellon University, Pittsburgh, PA 15213, USA
    index: 3
date: 28 February 2024
bibliography: paper.bib
---

# Summary

> The creation of healthy code is vital for its successful long term use in scientific research. To maximize impact throughout the community, software packages must be accurate, usable, and maintainable. Here we discuss several engineering processes that are important for developing healthy software. Unfortunately these processes often require configuration leading to short-term overhead for new projects. We introduce the LINCC-Frameworks Python Project Template, a configurable template designed for scientific software projects that greatly simplifies adopting such practices by automating the setup and configuration of important code health tools.

Reviewer 4: I am wondering what is meant by "accurate" (in "software packages must be accurate").  Is this a statement about numerical robustness of results, ...?  It read a bit oddly to me.  Possible alternate word that might resonate with science users: "robust"?

Author: I interpreted "accurate" as more closely related to "correct". Meaning that as the user makes changes, they can be confident that the unit tests will prevent breaking the code in such a way that it would produce incorrect output. 

I tend to think of "robust" as meaning "resilient" and will protect the user in case of odd inputs and provide meaningful error messages if the code can't recover on its own. However, if "robust" has a different connotation in the context of scientific software, I would be all for swapping in "robust" :) 

Do you have a strong opinion "robust" vs. "correct" vs. something else?

Author: Going with "robust" for now.

Author: Updated

Reviewer: I was just going to suggest either keeping "accurate" or going with "correct." The thinking was that it is important to make sure the code stays correct as changes are made. We've come across a few errors in various projects that could have been addressed by better testing.

Author: I'm also happy to go with "correct" :)

Reviewer 4: Hmm.  So let me explain the origin of my confusion in case helpful (sorry for not saying more before).  As a scientist, I know that in general most of my analysis algorithms that I implement have some approximations or limitations, and are not formally correct.  What I care about is that once I have implemented it with all of its warts, that the code accurately reflect my intentions.  I wouldn't say that such code is necessarily "accurate" or "correct" as it feels like there is an absoluteness to that statement that is a mismatch to the situation... do you see what you mean?  And maybe I'm just a bit of a weirdo in this regard, so it's OK if you want to ignore this feedback and move on.

Author: I believe I understand what you're saying @rmandelb, and it makes a lot of sense. To @jeremykubica point, the thought that we're trying to convey here with one word (that we haven't quite figured out yet) is "confident that there will be no regressions as changes are made". 

So perhaps instead of "correct", or "accurate", we could say, "stable" or "consistent"? Something like:
"software packages must be usable, maintainable, and produce consistent results"

Reviewer 4: I would be happy with that, thank you!  I'm OK if it's wrong for reasons I know/accept when I chose to implement a a particular algorithm.  It's the consistency with my intentions that matters.

# Statement of need

Software has long played a vital role in analyzing scientific data and driving new discoveries. As the scientific community continues to build upon established projects and create novel algorithms, there is a need to ensure the accuracy, repeatability, usability, and maintainability of the software. The first factor is required for the validity of the scientific results, while the latter factors enable the software to have a broad, sustained impact.

> The LINCC-Frameworks Python Project Template (LF-PPT) was originally created with the needs of astronomers in mind, however, it became apparent that it is broadly applicable for many scientific use cases. The LF-PPT codifies our best engineering practices because we were unable to find existing tooling that met our needs. Many other templates exist for specific applications that include non-trivial amounts of code, but we wanted a template that 1) is agnostic to specific applications 2) includes tooling needed for healthy software 3) doesn’t preclude use of other application-specific templates and 4) is updatable as best practices and tooling evolve. 

Reviewer 4: ```suggestion
The LINCC-Frameworks Python Project Template (LF-PPT) was originally created with the needs of astronomers in mind; however, it became apparent that it is broadly applicable for many scientific use cases. The LF-PPT codifies our best engineering practices because we were unable to find existing tooling that met our needs. Many other templates exist for specific applications that include non-trivial amounts of code, but we wanted a template that 1) is agnostic to specific applications, 2) includes tooling needed for healthy software, 3) does not preclude use of other application-specific templates, and 4) is updatable as best practices and tooling evolve. 
```

Reviewer 4: (just punctuation and "doesn't" -> "does not" for journal level of formality)

## Distribution

Software is only useful to the broader community if other users can find, install, and update it. Package management systems such as pip[^3] or conda[^4] provide simple tools for a user to download a new software package and its dependencies from public repositories such as PyPI[^5] or conda-forge[^6]. To enable easy code installation with pip or conda the LF-PPT provides support for automatic distribution through both PyPI and conda-forge when the user applies a new git tag via the GitHub UI.

## Additional code health tools

The LF-PPT supports many more features than those listed here including static code analysis, performance benchmarking, code testing coverage and more. For a complete list, please see the documentation at http://lincc-ppt.readthedocs.io/.

> The most notable exclusion from the LF-PPT is code. Aside from a few optional stub source and test files, the LF-PPT allows the user to focus on scientific code, while supporting them with the industry best practices in maintainable software engineering.

Reviewer 4: I found this paragraph a bit hard to understand (what is it trying to communicate and how does it relate to the "Additional code health tools" heading).  Can you elaborate on what you are trying to convey and how it relates to the rest of the subsection?   I mean we could maybe chat about it here in a comment in hopes of identifying potential edits to the text.

Author: Sure, good call here. What I want to convey is that this template provides the foundational pieces of a software package, but leaves the code (i.e. the science) to the user to implement. We want the template to lower the barrier to create a quality software package, without actually providing code.

I agree though that this can seem out of place. We don't call out that we provide code anywhere else, so I would be ok just cutting this paragraph. 

What are your thoughts?

Reviewer: I think it would be fine to cut the paragraph.

Reviewer 4: Hmm.  Thanks, so I think this reinforces my feeling that it doesn't belong here.  But I feel the idea that we are trying to make it easier for scientists to focus just on the code and not the infrastructure is a nice one and perhaps it could be worked (more concisely) into the summary at the top?

Author: Ok, I've cut this paragraph, and added a new line at the end of the summary paragraph:

"Notably the template does not include any application specific code. This allows the user to focus on their scientific code, and not building or maintaining code health infrastructure."

# Usage

## General usage

> The LINCC-Frameworks Python Project Template automates the setup of the above processes for Python projects hosted on GitHub. The only direct dependency is copier[^7] [@copier], which is used as the engine to generate new projects from the LF-PPT with a specific directory structure, the requested configuration and stub files. To hydrate a new or existing project with the LF-PPT via copier the user calls:

Reviewer 4: The word "hydrate" is being used here without being defined.  Are we sure readers will know what is meant?  (I only know because I've talked with you all...)

Author: Good call here, "hydrate" is a little too jargon-y. I'll change the wording here.

Author: Going with:
"To create a new project with the LF-PPT via copier, the user calls:"

> The LINCC-Frameworks Python Project Template automates the setup of the above processes for Python projects hosted on GitHub. The only direct dependency is copier[^7] [@copier], which is used as the engine to generate new projects from the LF-PPT with a specific directory structure, the requested configuration and stub files. To hydrate a new or existing project with the LF-PPT via copier the user calls:

Reviewer 4: ```suggestion
The LINCC-Frameworks Python Project Template automates the setup of the above processes for Python projects hosted on GitHub. The only direct dependency is copier[^7] [@copier], which is used as the engine to generate new projects from the LF-PPT with a specific directory structure, the requested configuration and stub files. To hydrate a new or existing project with the LF-PPT via copier, the user calls:
```

```
cd <new/project/directory>
bash .initialize-new-project.sh
```

The process is designed so as not to require significant time, and, thus, if the user is unhappy with the generated project, they can simply delete and recreate it. Depending on the options selected, some additional configuration may be required, such as registering with ReadTheDocs or PyPI. To assist the user, a customized, post-creation checklist is generated.

## Existing projects

In addition to creating new projects from scratch, the LF-PPT can be applied to existing projects to incorporate the features it provides. Often a more established project will require more effort to apply the LF-PPT. However, LINCC-Frameworks has had success applying the template to multiple legacy projects. A collection of tips for applying the template to existing projects is included in the documentation http://lincc-ppt.readthedocs.io/.

## Updating projects

> As best practices evolve and new tools are introduced, the LF-PPT will incorporate those into the template. Updates to the template can be applied to existing projects with minimal effort, allowing the users to focus on science and not software maintenance.

Reviewer 4: Is this as configurable as initial application of the template, where they get to decide which code health features they want to use?  If so, might be worth mentioning this configurability of updates.

Author: Yes, you're correct, `copier update` does allow the user to modify the initial choices. I'll add a line about that.

Author: Added this line after the :

"Similarly, if the user determines that different template options would better suit the project as it evolves, the same command can be used to review and change the initial responses to the questionnaire and apply those changes to the project."

```
copier update
```

## Hibernating projects

> Scientific projects often go into periods of hibernation when not under active development or use, and are often challenging to revive. With the LF-PPT, hibernating projects should be much easier to reactivate. Automatically scheduled smoke tests and dependabot[^8] integration ensure that the code and dependencies continue to function correctly without significant interaction from the maintainers.

Reviewer 4: ```suggestion
Scientific projects often go into periods of hibernation when not under active development or use, and can be challenging to revive. With the LF-PPT, hibernating projects should be much easier to reactivate. Automatically scheduled smoke tests and dependabot[^8] integration ensure that the code and dependencies continue to function correctly without significant intervention from the maintainers.
```

Reviewer 4: (the "often" / "often" duplication was bothering me; for the later thing, I think it should be either "intervention from" or "interaction with")

## Recent applications

> The LF-PPT has been applied to multiple LINCC Frameworks projects including LSDB[^9], Hipscat[^10], and TAPE[^11], short term collaborations such as Sorcha[^12], SuperPhot+[^13], DeepDISC[^14], and MacCauff[^15], and external project including kcorrect[^16] and FlexCode[^17] Additionally a project specific version was forked from the project and has been applied to all RAIL packages[^18].

Reviewer 4: ```suggestion
The LF-PPT has been applied to multiple LINCC Frameworks projects including LSDB[^9], HiPSCat[^10], and TAPE[^11], short term collaborations such as Sorcha[^12], SuperPhot+[^13], DeepDISC[^14], and MacCauff[^15], and external projects including kcorrect[^16] and FlexCode[^17]. Additionally, a project-specific version was forked from the project and has been applied to all RAIL packages[^18].
```

# Acknowledgments

> This project is supported by Schmidt Sciences.

Reviewer 4: I was curious whether there are any community members we might want to acknowledge - e.g. people who've submitted helpful feedback and feature requests, or who helped us advertise the work within LSST SCs, etc.?

Author: Let me think on this one a little bit, I'm sure there's a reasonable list that I can add here.

Reviewer 4: Yes, that's a good idea as well.
